cloud/scope/azureippool.go

```diff
@@ -0,0 +1,348 @@
+					if subnetName == allocatedIPPool.SubnetName {
+						foundAllocatedIPPool = true
+						allocatedIPPool, foundIPInAllocatedPool = RemoveIPFromAllocatedIPPool(allocatedIPPool, ip)
+						foundIP = foundIP || foundIPInAllocatedPool
+						if err!=nil {
+							return errors.Wrapf(err, "Failed to remove from allocated IP Pool")
+						}
+						azureIPPool.Status.AllocatedIPPools[index] = allocatedIPPool
+					} 
+				}
+				if !foundAllocatedIPPool {
+					return errors.New(fmt.Sprintf("Unable to find allocated IP pool for subnet name %s",subnetName)) 
+				}
+				if !foundIP {
+					return errors.New(fmt.Sprintf("Unable to find IP %s in allocated IP pools",ip))
+				}
+				machineScope.Info(fmt.Sprintf("Allocated IP pools is %v after removing IP %s", azureIPPool.Status.AllocatedIPPools, ip))
+			}
+		}
+
+		machineScope.Info(fmt.Sprintf("Allocated IP Pool is %v", azureIPPool.Status.AllocatedIPPools))
+		machineScope.Info(fmt.Sprintf("Azure IP Pool is %v", azureIPPool))
+		
+		//patch AzureIPPool
+		if err := s.patchHelper.Patch(ctx, azureIPPool); err != nil {
+			return errors.Wrapf(err, "Failed to update Azure IP pool")
+		}
+		machineScope.Info(fmt.Sprintf("Successfully updated AzureIPPool %s", azureIPPool.Name))
+
+		//resetting nic info on machine status
+		machineScope.Info(fmt.Sprintf("Assigning network interfaces to nil"))
+		machineScope.AzureMachine.Status.NetworkInterfaces = []infrav1.NetworkInterface{}
+
+		//resetting reconciledIP info on machine status
+		machineScope.Info(fmt.Sprintf("Setting ReconciledIP to false"))
+		machineScope.AzureMachine.Status.ReconciledIP = false
+
+		//patch azure machine
+		if err := machineScope.patchHelper.Patch(ctx, machineScope.AzureMachine); err != nil {
+			return errors.Wrapf(err, "Failed to update Azure Machine")
+		}
+		machineScope.Info(fmt.Sprintf("Successfully updated AzureMachine status"))	
+		
+	}				
+
+	return nil
+}
+
+func BuildNetworkInterfaceSpec(machineScope *MachineScope,ippool infrav1.IPPool, name string) infrav1.NetworkInterface {
+    networkInterface := infrav1.NetworkInterface{}
+    networkInterface.Name = azure.GenerateNICName(machineScope.Name())
+    networkInterface.VnetName = ippool.VnetName
+    networkInterface.SubnetName = ippool.SubnetName
+    networkInterface.VnetResourceGroup = "aserg"
+    if name == "mgmt-nic-ip-pool" {
+        networkInterface.IsPrimary = true
+    } else {
+		networkInterface.AcceleratedNetworking = true
+    }
+    return networkInterface
+}
+
+func GetPrimaryNetworkInterface(nics []infrav1.NetworkInterface) (infrav1.NetworkInterface, error) {
+    nic := infrav1.NetworkInterface{}
+	for _,nic := range(nics) {
+        if nic.IsPrimary == true {
+            return nic, nil
+        }
+    }
+    return nic, errors.New("Unable to find a nic with isPrimary set to true")
+}
+
+func AddToAllocatedIPPool(azureippool *infrav1.AzureIPPool, name string, allocatedIP string, vnetName string, subnetName string) *infrav1.AzureIPPool {
```

Comment:
Why not make this part of the AzureIPPoolScope.

Reply:
Does this need to be public?  I think it can be private.

Reply:
> Why not make this part of the AzureIPPoolScope.

Can you please explain what should be done here? Sorry did not get this.

Reply:
Change the function to:
func (s *AzureIPPoolScope) AddToAllocatedIPPool(name string, allocatedIP string, vnetName string, subnetName string) *infrav1.AzureIPPool)